### PR TITLE
refacctor fiter and unionall operator

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:
refactor filter operator and unionall operator

### CODE REVIEW

File: pkg/sql/colexec/filter/types.go [hunk 1]
```diff
@@ -79,15 +79,27 @@ func (filter *Filter) GetExeExpr() *plan.Expr {
 }
 
 func (filter *Filter) Reset(proc *process.Process, pipelineFailed bool, err error) {
-	filter.Free(proc, pipelineFailed, err)
+	//filter.Free(proc, pipelineFailed, err)
```
Comment (ouyuanning) [inline pkg/sql/colexec/filter/types.go:82]: 删掉这些注释吧

File: pkg/sql/colexec/filter/filter_test.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+)
+
+func init() {
+	boolType := types.T_bool.ToType()
+	int32Type := types.T_int32.ToType()
+
+	fr, _ := function.GetFunctionByName(nil, ">", []types.Type{int32Type, int32Type})
+	fid := fr.GetEncodedOverloadID()
+
+	tcs = []filterTestCase{
```
Comment (ouyuanning) [inline pkg/sql/colexec/filter/filter_test.go:52]: 可以加一些testcase，可以包含以下情况
1、只有1个筛选条件
2、有多个筛选条件，每次能筛选一部分
3、子节点有2个batch的情况

File: pkg/sql/colexec/filter/filter.go [hunk 3]
```diff
@@ -79,13 +73,20 @@ func (filter *Filter) Call(proc *process.Process) (vm.CallResult, error) {
...
 	}
-	if filter.ctr.buf != nil {
-		proc.PutBatch(filter.ctr.buf)
-		filter.ctr.buf = nil
-	}
-	filter.ctr.buf = result.Batch
 
-	anal.Input(filter.ctr.buf, filter.GetIsFirst())
+	if filter.ctr.buf == nil {
+		newBat, err := result.Batch.Dup(proc.Mp())
```
Comment (fengttt) [inline pkg/sql/colexec/filter/filter.go:78]: I think this code is *correct* but not *efficient*

Both if and else branch, we basically copy the batch.   A much better approach is to run the first filter, and copy the filter result into filter.ctr.buf

File: pkg/sql/colexec/filter/filter_test.go (hunk 2 above)
Comment (qingxinhome) [reply pkg/sql/colexec/filter/filter_test.go:58]: 已经补充测试用例，请review

File: pkg/sql/colexec/filter/types.go (hunk 1 above)
Comment (qingxinhome) [reply pkg/sql/colexec/filter/types.go]: OK

File: pkg/sql/colexec/filter/filter.go (hunk 3 above)
Comment (qingxinhome) [reply pkg/sql/colexec/filter/filter.go:78]: 这种方式有点复杂，而且能节省的条件比较苛刻，要不暂时先这样做，看看效果如何

Comment (fengttt) [reply pkg/sql/colexec/filter/filter.go:78]: I think this is critical for performance.

**Review by fengttt — COMMENTED** [review]
I think the filter copy is critical for performance.  Should do the right thing here.

**Comment by qingxinhome** [comment]
> I think the filter copy is critical for performance. Should do the right thing here.
> 
> I don't mind you submit this, then post another PR.

OK, Optimized, please review @fengttt

**Pull request merged**